webapp/src/components/backend_selector.jsx

```diff
@@ -0,0 +1,137 @@
+
+export default class BackendSelector extends PureComponent {
+    static propTypes = {
+        name: PropTypes.string,
+        required: PropTypes.bool,
+        isMulti: PropTypes.bool,
+        resetInvalidOnChange: PropTypes.bool,
+        addValidate: PropTypes.func,
+        removeValidate: PropTypes.func,
+        onChange: PropTypes.func,
+        load: PropTypes.func,
+        theme: PropTypes.object.isRequired,
+        value: PropTypes.oneOfType([
+            PropTypes.object,
+            PropTypes.array,
+            PropTypes.string,
+        ]),
+    };
+
+    constructor(props) {
+        super(props);
+
+        this.state = {
+            options: [],
+            invalid: false,
+            error: '',
+        };
+    }
+
+    componentDidMount() {
+        if (this.props.addValidate) {
+            this.props.addValidate(this.isValid);
+        }
+    }
+
+    componentWillUnmount() {
+        if (this.props.removeValidate) {
+            this.props.removeValidate(this.isValid);
+        }
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (prevState.invalid && this.props.value !== prevProps.value) {
+            this.setState({invalid: false}); //eslint-disable-line react/no-did-update-set-state
+        }
+    }
+
+    onChange = (options) => {
+        if (!options) {
+            this.props.onChange(this.props.isMulti ? [] : '');
+            return;
+        }
+
+        this.props.onChange(this.props.isMulti ? options.map((v) => v.value) : options.value);
+
+        if (this.props.resetInvalidOnChange) {
+            this.setState({invalid: false});
+        }
+    };
+
+    isValid = () => {
+        if (!this.props.required) {
+            return true;
+        }
+
+        const valid = Boolean(this.props.value && this.props.value.toString().length !== 0);
+        this.setState({invalid: !valid});
+        return valid;
+    };
+
+    errorComponent = () => {
+        if (!this.state.error) {
+            return null;
+        }
+
+        return (
+            <p className='alert alert-danger'>
+                <i
+                    className='fa fa-warning'
+                    title='Warning Icon'
+                />
+                <span> {this.state.error.toString()}</span>
+            </p>
+        );
+    };
+
+    validationError = () => {
+        if (!this.props.required || !this.state.invalid) {
+            return null;
+        }
+
+        return (
+            <p className='help-text error-text'>
+                <span>{'This field is required.'}</span>
+            </p>
+        );
+    };
+
+    loadOptions = async () => {
+        const options = await this.props.load();
+
+        // prevent re-render if the options remain unchanged
+        if (JSON.stringify(options) === JSON.stringify(this.state.options)) {
+            return;
+        }
+
+        this.setState({options});
+    }
+
+    render = () => {
+        this.loadOptions();
```

Comment:
This currently calls `this.loadOptions` on every render, which doesn't seem like something we want to do. We instead would want to load these in `componentDidMount` since they will only be loaded once. This is taken care of in the `BackendSelector` component from the Jira plugin.

Reply:
@mickmister The reason I had to call this on every render was that we have to reload the labels whenever the selected repository changes. 

More specifically, `GithubLabelSelector` re-renders when the selected repository changes, which also causes a re-render for this component. If I don't call `loadOptions` within `render`, our list of labels will become stale. 

That said, I'm definitely open to suggestions on how we can make this more efficient.

Reply:
We can instead use the [componentDidUpdate](https://reactjs.org/docs/react-component.html#componentdidupdate) method, and compare the previous repo value with the current one being passed in. If the value has changed, then we re-fetch the options.